manuals/sigscan.1
.Dd June 27, 2026
.Dt SIGSCAN 1
.Os
.Sh NAME
.Nm sigscan
.Nd scans a file for binary signatures
.Sh SYNOPSIS
.Nm sigscan
.Op Fl c Ar path
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm sigscan
is a utility to scan a file for binary signatures
.Pp
.Nm sigscan
is part of the
.Nm libsigscan
package.
.Nm libsigscan
is a library to scan a file for binary signatures
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar path
specify the configuration file, defaults to: sigscan.conf
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# sigscan -c sigscan.conf qcow.bin
sigscan 20260627
.sp
Signature scanner:
	Number of scan results	: 1
.sp
Scan result: 1
	Identifier		: qcow
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libsigscan/issues
.Sh COPYRIGHT
Copyright (C) 2014-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* buffer scan
  - align with footer range
* unbounded scan
  - determine optimal buffer size
  - buffer size vs scan block size?
  - handle boundary matches
  - add range to determine size of scan buffer?

* patterns support
  - is there a pratical need to support signature with patterns that contain pattern gaps?
* make sure scan buffer is sufficient large for bounded scan
* state add relative_from_start, relative_from_end and unbound scan trees and active nodes
* add memory aligned reads?

Tools:
* sigscan/scan_handle
  - print result offset

Tests:
* add test signature files
  - tar, tar_old
* sigscan_test_scanner
  - add scan result (signature) identifier check
  - offsets relative from end
  - scan file and file IO handle
  - build scan tree with no signatures, single signature, multiple signatures
  - different buffer sizes
  - header signature range that does not start at 0
* pysigscan_test_scanner
  - fix TODOs
  - offsets relative from end
  - scan file and file object
  - build scan tree with no signatures, single signature, multiple signatures
* sigscan
  - add test

20141212
* see `git log' for more recent change log
* worked on initial version

20141208
* worked on initial version

20141116
* worked on initial version

20141115
* Initial version based on dfvfs scanner and libexe



sigscantools/sigscan.c
/*
 * Scans a file for binary signatures
 *
 * Copyright (C) 2014-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "scan_handle.h"
#include "sigscantools_getopt.h"
#include "sigscantools_libcerror.h"
#include "sigscantools_libclocale.h"
#include "sigscantools_libcnotify.h"
#include "sigscantools_libsigscan.h"
#include "sigscantools_output.h"
#include "sigscantools_signal.h"
#include "sigscantools_unused.h"

scan_handle_t *sigscan_scan_handle = NULL;
int sigscan_abort                  = 0;

/* Signal handler for sigscan
 */
void sigscan_signal_handler(
      sigscantools_signal_t signal SIGSCANTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "sigscan_signal_handler";

	SIGSCANTOOLS_UNREFERENCED_PARAMETER( signal )

	sigscan_abort = 1;

	if( sigscan_scan_handle != NULL )
	{
		if( scan_handle_signal_abort(
		     sigscan_scan_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use sigscan to scan a file for binary signatures.";

	sigscantools_option_t options[ ] = {
		{ 'c', "path", "specify the configuration file, defaults to: sigscan.conf" },
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source image" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error                      = NULL;
	system_character_t *option_configuration_file = _SYSTEM_STRING( "sigscan.conf" );
	system_character_t *source                    = NULL;
	libsigscan_scan_state_t *scan_state           = NULL;
	char *program                                 = "sigscan";
	system_integer_t option                       = 0;
	int number_of_options                         = (int) ( sizeof( options ) / sizeof( sigscantools_option_t ) );
	int verbose                                   = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "sigscantools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( sigscantools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	sigscanoutput_version_fprint(
	 stdout,
	 program );

	if( sigscantools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = sigscantools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				sigscantools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_configuration_file = optarg;

				break;

			case (system_integer_t) 'h':
				sigscantools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				sigscanoutput_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		sigscantools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libsigscan_notify_set_stream(
	 stderr,
	 NULL );
	libsigscan_notify_set_verbose(
	 verbose );

/* TODO check if option_configuration_file exists */

	if( scan_handle_initialize(
	     &sigscan_scan_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to create scan handle.\n" );

		goto on_error;
	}
	if( libsigscan_scan_state_initialize(
	     &scan_state,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to create scan state.\n" );

		goto on_error;
	}
	if( scan_handle_read_signature_definitions(
	     sigscan_scan_handle,
	     option_configuration_file,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to read signatures from: %" PRIs_SYSTEM ".\n",
		 option_configuration_file );

		goto on_error;
	}
	if( scan_handle_scan_input(
	     sigscan_scan_handle,
	     scan_state,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to scan: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( libsigscan_scan_state_free(
	     &scan_state,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free scan state.\n" );

		goto on_error;
	}
	if( scan_handle_free(
	     &sigscan_scan_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( scan_state != NULL )
	{
		libsigscan_scan_state_free(
		 &scan_state,
		 NULL );
	}
	if( sigscan_scan_handle != NULL )
	{
		scan_handle_free(
		 &sigscan_scan_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

